### ISSUE TITLE
Correctly load similar clips for fine-tuning

### ISSUE THREAD

**Issue by 152334H** [152334H/DL-Art-School#11]
# PROBLEM

Let's start from the [`EXAMPLE_gpt.yml`](https://github.com/152334H/DL-Art-School/blob/master/experiments/EXAMPLE_gpt.yml) file.

There's a parameter named `num_conditioning_candidates` under `datasets.train` && `datasets.val`. It (ostensibly) determines the number of conditioning wav files that are piped to the gpt model during training.

How does it work? In [`paired_voice_audio_dataset.py`](https://github.com/152334H/DL-Art-School/blob/master/codes/data/audio/paired_voice_audio_dataset.py#L200-L201), there's a line in `__getitem__` that grabs `num_conditioning_candidates` similar clips from the current dataset:

```python
            cond, cond_is_self = load_similar_clips(self.audiopaths_and_text[index][0], self.conditioning_length, self.sample_rate,
                                      n=self.conditioning_candidates) if self.load_conditioning else (None, False)
```
Or at least, that's what it's supposed to in theory. In practice, [it **ignores the value** of `n` passed](https://github.com/152334H/DL-Art-School/blob/master/codes/data/audio/unsupervised_audio_dataset.py#L49-L62), because of the lack of a file named `similarities.pth`:

```python
def load_similar_clips(path, sample_length, sample_rate, n=3, fallback_to_self=True):
    sim_path = os.path.join(os.path.dirname(path), 'similarities.pth')
    candidates = []
    if os.path.exists(sim_path): # obviously ignored
    if len(candidates) == 0: # true
        if fallback_to_self: # true
            candidates = [path] # ONLY 1 CANDIDATE USED
    #...
```

(i also printed out the values for `cond_is_self` and verified that it's true right now)

The `similarities.pth` file(s?) are supposed to be generated by a script titled [`phase_3_generate_similarities.py`](https://github.com/152334H/DL-Art-School/blob/master/codes/scripts/audio/preparation/phase_3_generate_similarities.py#L50). I saw the preparation scripts earlier, but to be real I have no idea how to use them yet.

# why is this a problem?

Plausibly could lead to the [cheater latents](https://nonint.com/2022/06/18/lab-notes-cheater-latents/) problem on larger datasets. Also probably reduces the zero shot vc capabilities.

**Comment by 152334H** [comment]
Okay, I tried running it. Conclusion: not possible.

## Phase 1
Simple enough to run. There is a problem with the script exploding if you provide it relative paths, so just use absolute paths (`readlink -f`) for everything.

It's not complicated anyway, it just splits audio files that are too long.

## Phase 2
This second phase is supposed to make use of a basic classifier that checks for noisy clips.

### Problem: classifier model is missing

I have no idea where the pre-trained classifier model is supposed to be. Looking at the model code files, it seems likely that it's some kind of basic resnet deal, so that could plausibly be retrained. But I'm not willing to do that atm.

## Phase 3
As stated above, this script is supposed to create similarity metrics for audio files in the dataset.

### Problem: voice_voice_clip model is missing

> "Wait, are you sure? The tortoise-tts repo has a CVVP model, right? Couldn't you rank similarities with that?"

Apparently not. The CVVP model is conditioned on a (`mel_input`, `mel_cond`) pair, the VoiceCLIP model takes (`speech_mels`, `speech_lengths`) as input. Also, their architectures are wildly different.

---

In short, 2/3 phases are screwed, and the first one is mostly useless. This issue will have to be tossed to the back of the TODO until much later.
